React: Add Introduction to State lesson (#25261)
TheOdinProject/curriculum: react/states_and_effects/introduction_to_state.md

### Introduction

> Any exciting application you build is likely to change over its lifetime. The change could be as simple as toggling a dropdown menu or as complex as fetching data from an API. React provides primitives to manipulate state of your apps, more specifically components, to make them dynamic. In this lesson, we will learn how to use state in React.  

Reviewer: `1.` lifetime. __The change__ could be `to` lifetime. __Changes__ could be
`2.` to manipulate __state__ of your `to` to manipulate __the state__ of

Read  [What is State? by Academind](https://academind.com/tutorials/what-is-state) before we get started.

### Lesson Overview

This section contains a general overview of topics that you will learn in this lesson.

* What is state?
* How to use state in React.   

> * What happens when state changes in React?

Reviewer: when __state__ changes in `to` when __the state__ changes in

### What is State in React? 

We write components in React, A LOT of components, and many times we want them to undergo visual changes as a result of user or computer interactions. For this purpose, a component needs to "remember" things about itself. This is where state comes in. **State is a component's memory.** 

> Let's take a look on how we can use state in our code.

Reviewer: a look __on__ how we can `to` a look __at__ how we can

```
CODESANDBOX EXAMPLE GOES HERE
```

An app that changes its background color based on the clicked button. Pretty simple, right? Take your time reading the code as we'll jump right in.

#### The useState hook

> <span id="useState-hook">The `useState` hook is a built-in hook (we'll talk about hooks later) in React that allows you to define state in a functional component. It takes an initial value as a parameter and returns an array with two elements:</span>

Reviewer: with two __elements:__ `to` with two __elements that we can destucture to get:__

1. The current state value
1. A function to update the state value

State definition with `useState` commonly follows this pattern:

~~~js
const [stateValue, setStateValue] = useState(initialValue);

// adapted for our use case:
const [backgroundColor, setBackgroundColor] = useState(initialColor);
~~~

Even without much knowledge of React, you can, to some extent, understand what's going on. The `backgroundColor` state is defined with the hook. Then on every button, we set up a _click_ event handler that calls the `setBackgroundColor` function with the corresponding value. Then, magically the new color is applied to the background. 

Adding more state variables should be as easy as adding more `useState` calls.

> Great, you can now use state! But, what's actually going on under the hood? 

Reviewer: But, what's __actually going__ on `to` But, what's __going__ on

### How does State Work in React? 

Let's hit you with some theory.

In React, when a component's state or props change, the component is destroyed and recreated from scratch. Yes, you heard that right: _destroyed_. This includes the variables, functions, and React nodes. The entire component is recreated but this time the latest state value will be returned from `useState`. This process is called rerendering. Rerendering is a key feature of React that enables it to efficiently update the user interface in response to changes in the underlying data.

<div class="lesson-note" markdown="1">

#### React Reconciliation Algorithm

The process of rerendering generates a new virtual DOM (Document Object Model) tree. The virtual DOM is a lightweight representation of the actual DOM that React uses to keep track of the current state of the UI. React then compares the new virtual DOM tree to the previous one and calculates the minimal set of changes needed to update the actual DOM. This is the reconciliation algorithm. 

</div>

#### Explaining rerendering with the example above 

> Whenever `setBackgroundColor` is called, our `App` component is rerendered. Essentially, the component is recreated which means the `onButtonClick` function and our `div` and `button`'s are recreated as well. You might wonder, shouldn't the `backgroundColor` state be recreated as well? Well, React takes the responsibility of keeping track of the latest state and providing it to the component. The initial state value is only used for the component's first render, and is ignored on subsequent renders. 

Reviewer: component's first __render,__ and is `to` component's first __render__ and is

![code illustration of rerendering](./introduction_to_state/imgs/00.png)

What we covered in this section is just a brief overview of a portion of React internals. You may be curious to learn more, but be advised, this is a deep rabbit hole. What we covered above is enough to get you through the React course. 

### Hooks 

> Hooks are functions that let you use React features. All hooks are recognizable by the `use` prefix. For example, `useState` is a hook. We'll use more of these as we get further into the course. For now, remember that hooks have rules that we need to abide by:

Author: I'm not sure what you mean by PascalCase, aren't hooks conventionally camelCase?

> Hooks are functions that let you use React features. All hooks are recognizable by the `use` prefix. For example, `useState` is a hook. We'll use more of these as we get further into the course. For now, remember that hooks have rules that we need to abide by:

Reviewer: My bad, it should be camelCase.

> Hooks are functions that let you use React features. All hooks are recognizable by the `use` prefix. For example, `useState` is a hook. We'll use more of these as we get further into the course. For now, remember that hooks have rules that we need to abide by:

Author: Learners should already be plenty familiar with camelCase i.e. they should be able to identify that hooks, say `useState` follow camelCase convention. Based on what I said, what if we skip mentioning hooks are camelCase here?

> Hooks are functions that let you use React features. All hooks are recognizable by the `use` prefix. For example, `useState` is a hook. We'll use more of these as we get further into the course. For now, remember that hooks have rules that we need to abide by:

Reviewer: Looks like that is covered in the assignments too, yeah, sounds fair.

1. Hooks can only be called from the top level of a functional component.
1. Hooks can't be called from inside loops or conditions.

### Assignment

<div class="lesson-content__panel" markdown="1">

> 1. Read [State: A Component's Memory from the React docs](https://react.dev/learn/state-a-components-memory).

Reviewer: I feel like all of the first three of the react docs sections on `Adding interactivity` are discussed in this lesson. So maybe including read throughs of the other two (`Render and commit` and `State as a Snapshot`) may be good decision?

1. Read this [article on React Reconciliation Algorithm](https://medium.com/javarevisited/react-reconciliation-algorithm-86e3e22c1b40) for a great explanation.
1. Add a new state variable to keep track of the number of times the background color has been changed. Display the number of times the background color has been changed on the page. 

</div>

### Knowledge Check

This section contains questions for you to check your understanding of this lesson on your own. If you’re having trouble answering a question, click it and review the material it links to.

*   <a class="knowledge-check-link" href="#what-is-state-in-react">What is state?</a>
*   <a class="knowledge-check-link" href="#useState-hook">What is the `useState` hook and how to use it?</a>
*   <a class="knowledge-check-link" href="#how-does-state-work-in-react">What happens to the component when one of its state is changed?</a>

> *   <a class="knowledge-check-link" href="#hooks">What are the rules of hooks?</a>

Reviewer: are __the__ rules `to` are __some of the__ rules

### Introduction

> Any exciting application you build is likely to change over its lifetime. The changes could be as simple as toggling a dropdown menu or as complex as fetching data from an API. React provides primitives to manipulate the state of your apps, more specifically components, to make them dynamic. In this lesson, we will learn how to use state in React.  

Reviewer: ```suggestion
Any exciting application you build is likely to change over the time the user is exploring it. The changes could be as simple as toggling a dropdown menu or as complex as fetching data from an API. React provides primitives to manipulate the state of your apps, more specifically components, to make them dynamic. In this lesson, we will learn how to use state in React.  
```

In this initial line, I'm not sure how I feel about the wording. "change over lifetime" can refer to many things including the developmental changes, updates, etc.

Author: New wording looks good!

![code illustration of rerendering](./introduction_to_state/imgs/00.png)

What we covered in this section is just a brief overview of a portion of React internals. You may be curious to learn more, but be advised, this is a deep rabbit hole. What we covered above is enough to get you through the React course. 

### Hooks 

Hooks are functions that let you use React features. All hooks are recognizable by the `use` prefix. For example, `useState` is a hook. We'll use more of these as we get further into the course. For now, remember that hooks have rules that we need to abide by:

1. Hooks can only be called from the top level of a functional component.
1. Hooks can't be called from inside loops or conditions.

### Assignment

<div class="lesson-content__panel" markdown="1">

1. Read the following articles from the React docs:
    1. [State: A Component's Memory](https://react.dev/learn/state-a-components-memory)

>     1. [Render and Commit](https://react.dev/learn/render-and-commit)

Author: @Lofty-Brambles  Agreed on Render and Commit article. I'm not sure on adding the State as a Snapshot article here as I'll be talking about state updater functions in the next lesson. I think State as a Snapshot will fit there well. What do you think?

Reviewer: Sounds like a good call to me. I wasn't too clear on the next lesson's objectives.
